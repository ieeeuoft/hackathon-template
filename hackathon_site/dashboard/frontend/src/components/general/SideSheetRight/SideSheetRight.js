import React from "react";
import Drawer from "@material-ui/core/Drawer";
import Typography from "@material-ui/core/Typography";
import Link from "@material-ui/core/Link";
import Button from "@material-ui/core/Button";
import MenuItem from "@material-ui/core/MenuItem";
import FormControl from "@material-ui/core/FormControl";
import Select from "@material-ui/core/Select";
import LaunchIcon from "@material-ui/icons/Launch";
import ArrowBackIcon from "@material-ui/icons/ArrowBack";
import FiberManualRecordIcon from "@material-ui/icons/FiberManualRecord";

import styles from "./SideSheetRight.module.scss";

const createQuantityList = (number, handleChange) => {
    let entry = [];

    for (let i = 1; i <= number; i++) {
        entry.push(
            <MenuItem onClick={() => handleChange(i)} key={i} role="quantity">
                {i}
            </MenuItem>
        );
    }

    return entry;
};

const CartSheet = ({ addCartFunction, quantity }) => {
    const [item, setItem] = React.useState("");

    const handleChange = (value) => {
        setItem(value);
    };

    return (
        addCartFunction && (
            <div className={styles.cart}>
                <FormControl className={styles.form}>
                    <Select value={item} role="selected">
                        {createQuantityList(quantity, handleChange)}
                    </Select>
                </FormControl>
                <Button
                    variant="contained"
                    className={styles.cartButton}
                    onClick={addCartFunction}
                    disableElevation
                >
                    ADD TO CART
                </Button>
            </div>
        )
    );
};

const BodySheet = ({ manufacturer, model_num, datasheet, notes, constraints }) => {
    return (
        <div className={styles.bodysheet}>
            <div className={styles.bodyinfo}>
                <Typography variant="h3">Manufacturer</Typography>
                <Typography>{manufacturer}</Typography>
            </div>
            <div className={styles.bodyinfo}>
                <Typography variant="h3">Model Number</Typography>
                <Typography>{model_num}</Typography>
            </div>
            <div className={styles.bodyinfo}>
                <Typography variant="h3">Datasheet</Typography>
                <Link
                    href={datasheet}
                    rel="noopener"
                    color="inherit"
                    underline="none"
                    target="_blank"
                    className={styles.bodyinfoDataSheet}
                >
                    <LaunchIcon></LaunchIcon>
                    <Typography>Link</Typography>
                </Link>
            </div>
            <div className={styles.bodyinfo}>
                <Typography variant="h3">Notes</Typography>
                {notes.map((note, i) => (
                    <Typography key={i}>{note}</Typography>
                ))}
            </div>
            <div className={styles.bodyinfo}>
                <Typography variant="h3">Constraints</Typography>
                {constraints.map((constraint, i) => (
                    <Typography key={i}>{constraint}</Typography>
                ))}
            </div>
        </div>
    );
};

const HeaderSheet = ({ type, name, total, available, tags, img }) => {
    return (
        <div className={styles.headersheet}>
            <div>
                <div className={styles.title}>
                    <FiberManualRecordIcon style={{ color: type, margin: "0 5px" }} />
                    <Typography variant="h2">{name}</Typography>
                </div>
                <Typography className={styles.available}>
                    {available} OF {total} IN STOCK
                </Typography>
                <div className={styles.tags}>
                    <Typography variant="h3">Tags</Typography>
                    <Typography>{tags}</Typography>
                </div>
            </div>
            <img src={img} alt="product" />
        </div>
    );
};

const SideSheetRight = ({ detail, addCartFunction }) => {
    const [toggle, setToggle] = React.useState(false);

    const toggleSheet = (change) => {
        setToggle(change);
    };

    return (
        <div>
            <div
                onClick={() => {
                    toggleSheet(true);
                }}
            >
                click me
            </div>
            <Drawer anchor="right" open={toggle}>
                <div className={styles.topsheet}>
                    <ArrowBackIcon
                        className={styles.backarrow}
                        onClick={() => {
                            toggleSheet(false);
                        }}
                        role="close"
                    />
                    <Typography variant="h2" className={styles.title}>
                        Product Overview
                    </Typography>
                </div>
                <HeaderSheet
                    type={detail.type}
                    name={detail.name}
                    total={detail.total}
                    available={detail.available}
                    tags={detail.tags}
                    img={detail.img}
                />
                <BodySheet
                    manufacturer={detail.manufacturer}
                    model_num={detail.model_num}
                    datasheet={detail.datasheet}
                    notes={detail.notes}
                    constraints={detail.constraints}
                />
                <CartSheet
                    addCartFunction={addCartFunction}
                    quantity={detail.quantity}
                />
            </Drawer>
        </div>
    );
};

export default SideSheetRight;
