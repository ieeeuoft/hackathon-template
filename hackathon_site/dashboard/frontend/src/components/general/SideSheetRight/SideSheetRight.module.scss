@import "./../../../assets/abstracts/mixins";
@import "./../../../assets/abstracts/variables";

.header {
    @include flexPosition($alIte: center);
    width: 100%;
    background-color: color(light2);
    z-index: 5;
    padding: 8px 34px;

    @include responsive(sm-down) {
        padding: 4px 12px;
    }

    &BackArrow {
        margin-right: 5px;
    }

    > h2 {
        @include responsive(md-down) {
            font-size: size(md) !important;
        }
    }
}

.body {
    width: 100%;
    height: 100%;
    padding: 25px 30px;

    @include responsive(sm-up) {
        padding: 45px 50px;
        min-width: 550px;
    }
}
