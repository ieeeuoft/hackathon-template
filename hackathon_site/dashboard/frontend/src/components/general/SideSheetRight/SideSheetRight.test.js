import React from "react";
import SideSheetRight from "./SideSheetRight";
import {
    render,
    fireEvent,
    waitForElementToBeRemoved,
    waitFor,
} from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";

/**
 * Test the quantity log thing
 */

describe("SideSheetRight", () => {
    const testDetail = {
        name: "Arduino",
        type: "red",
        total: 30,
        available: 29,
        img: "https://www.filterforge.com/more/help/images/size200.jpg",
        tags: "MCU, FPGA",
        manufacturer: "Canakit",
        model_num: "Model 3B+",
        datasheet: "link",
        notes: ["- For micropython ask for image", "- randomnerdtutorials.com"],
        constraints: ["- Max 1 of this item", "- Max 3 microcontroller labelled red"],
        quantity: 3,
    };

    const addCartMock = jest.fn();

    test("render", () => {
        const { asFragment } = render(
            <SideSheetRight detail={testDetail} addCartFunction={addCartMock} />
        );

        expect(asFragment()).toMatchSnapshot();
    });

    test("click me opens the drawer", () => {
        const { getByText, queryByText } = render(
            <SideSheetRight detail={testDetail} addCartFunction={addCartMock} />
        );

        //Open Drawer
        const openDrawer = getByText("click me");
        fireEvent.click(openDrawer);

        // Confirm open drawer
        expect(queryByText("Product Overview")).not.toBeNull();
    });

    test("close the drawer works", async () => {
        const { getByText, getByRole } = render(
            <SideSheetRight detail={testDetail} addCartFunction={addCartMock} />
        );
        //Open Drawer
        const openDrawer = getByText("click me");
        fireEvent.click(openDrawer);

        await waitFor(() => {
            expect(getByRole("close")).toBeInTheDocument();
        });

        const returnArrow = getByRole("close");
        fireEvent.click(returnArrow);

        await waitForElementToBeRemoved(() => getByText("Product Overview"));
    });

    test("add to cart button calls the correct function", () => {
        const { getByText } = render(
            <SideSheetRight detail={testDetail} addCartFunction={addCartMock} />
        );

        // open the drawer
        const openDrawer = getByText("click me");
        fireEvent.click(openDrawer);

        const button = getByText("ADD TO CART");

        fireEvent.click(button);
        expect(addCartMock).toHaveBeenCalledTimes(1);
    });
});
