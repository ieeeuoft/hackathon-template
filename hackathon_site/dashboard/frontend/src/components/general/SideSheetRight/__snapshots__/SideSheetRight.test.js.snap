// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideSheetRight render 1`] = `
<DocumentFragment>
  <div>
    <div>
      click me
    </div>
  </div>
</DocumentFragment>
`;
